Nama: Nicholas Darren Rafael

NIM: 2702261012

Model Deployment Mid Exam

# Import Libraries

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import pickle
import os
import joblib
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Data Preparation

In [3]:
df = pd.read_csv('Dataset_B_hotel.csv', sep =',')
df.head(10)

Booking_ID  no_of_adults  no_of_children  no_of_weekend_nights  \
0   INN00001             2               0                     1   
1   INN00002             2               0                     2   
2   INN00003             1               0                     2   
3   INN00004             2               0                     0   
4   INN00005             2               0                     1   
5   INN00006             2               0                     0   
6   INN00007             2               0                     1   
7   INN00008             2               0                     1   
8   INN00009             3               0                     0   
9   INN00010             2               0                     0   

   no_of_week_nights type_of_meal_plan  required_car_parking_space  \
0                  2       Meal Plan 1                         0.0   
1                  3      Not Selected                         0.0   
2                  1       Meal Plan 1                         0.0   
3                  2       Meal Plan 1                         0.0   
4                  1      Not Selected                         0.0   
5                  2       Meal Plan 2                         0.0   
6                  3       Meal Plan 1                         0.0   
7                  3       Meal Plan 1                         0.0   
8                  4       Meal Plan 1                         0.0   
9                  5       Meal Plan 1                         0.0   

  room_type_reserved  lead_time  arrival_year  arrival_month  arrival_date  \
0        Room_Type 1        224          2017             10             2   
1        Room_Type 1          5          2018             11             6   
2        Room_Type 1          1          2018              2            28   
3        Room_Type 1        211          2018              5            20   
4        Room_Type 1         48          2018              4            11   
5        Room_Type 1        346          2018              9            13   
6        Room_Type 1         34          2017             10            15   
7        Room_Type 4         83          2018             12            26   
8        Room_Type 1        121          2018              7             6   
9        Room_Type 4         44          2018             10            18   

  market_segment_type  repeated_guest  no_of_previous_cancellations  \
0             Offline               0                             0   
1              Online               0                             0   
2              Online               0                             0   
3              Online               0                             0   
4              Online               0                             0   
5              Online               0                             0   
6              Online               0                             0   
7              Online               0                             0   
8             Offline               0                             0   
9              Online               0                             0   

   no_of_previous_bookings_not_canceled  avg_price_per_room  \
0                                     0               65.00   
1                                     0              106.68   
2                                     0               60.00   
3                                     0              100.00   
4                                     0               94.50   
5                                     0              115.00   
6                                     0                 NaN   
7                                     0              105.61   
8                                     0               96.90   
9                                     0              133.44   

   no_of_special_requests booking_status  
0                       0   Not_Canceled  
1                       1   Not_Canceled  
2                 

In [4]:
df.tail(10)

Booking_ID  no_of_adults  no_of_children  no_of_weekend_nights  \
36265   INN36266             2               0                     1   
36266   INN36267             2               0                     2   
36267   INN36268             2               0                     1   
36268   INN36269             1               0                     0   
36269   INN36270             2               2                     0   
36270   INN36271             3               0                     2   
36271   INN36272             2               0                     1   
36272   INN36273             2               0                     2   
36273   INN36274             2               0                     0   
36274   INN36275             2               0                     1   

       no_of_week_nights type_of_meal_plan  required_car_parking_space  \
36265                  3       Meal Plan 1                         0.0   
36266                  2       Meal Plan 1                         0.0   
36267                  0      Not Selected                         0.0   
36268                  3       Meal Plan 1                         0.0   
36269                  1       Meal Plan 1                         0.0   
36270                  6       Meal Plan 1                         0.0   
36271                  3       Meal Plan 1                         0.0   
36272                  6       Meal Plan 1                         0.0   
36273                  3      Not Selected                         0.0   
36274                  2       Meal Plan 1                         0.0   

      room_type_reserved  lead_time  arrival_year  arrival_month  \
36265        Room_Type 1         15          2018              5   
36266        Room_Type 2          8          2018              3   
36267        Room_Type 1         49          2018              7   
36268        Room_Type 1        166          2018             11   
36269        Room_Type 6          0          2018             10   
36270        Room_Type 4         85          2018              8   
36271        Room_Type 1        228          2018             10   
36272        Room_Type 1        148          2018              7   
36273        Room_Type 1         63          2018              4   
36274        Room_Type 1        207          2018             12   

       arrival_date market_segment_type  repeated_guest  \
36265            30              Online               0   
36266             4              Online               0   
36267            11              Online               0   
36268             1             Offline               0   
36269             6              Online               0   
36270             3              Online               0   
36271            17              Online               0   
36272             1              Online               0   
36273            21              Online               0   
36274            30             Offline               0   

       no_of_previous_cancellations  no_of_previous_bookings_not_canceled  \
36265                             0                                     0   
36266                             0                                     0   
36267                             0                                     0   
36268                             0                                     0   
36269                             0                                     0   
36270                             0                                     0   
36271                             0                                     0   
36272                             0                                     0   
36273                             0                                     0   
36274                             0                                     0   

       avg_price_per_room  no_of_special_requests booking_status  
36265              100.73                       0   Not_Canceled  
36266            

In [5]:
df.shape

(36275, 19)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36275 entries, 0 to 36274
Data columns (total 19 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Booking_ID                            36275 non-null  object 
 1   no_of_adults                          36275 non-null  int64  
 2   no_of_children                        36275 non-null  int64  
 3   no_of_weekend_nights                  36275 non-null  int64  
 4   no_of_week_nights                     36275 non-null  int64  
 5   type_of_meal_plan                     35368 non-null  object 
 6   required_car_parking_space            35005 non-null  float64
 7   room_type_reserved                    36275 non-null  object 
 8   lead_time                             36275 non-null  int64  
 9   arrival_year                          36275 non-null  int64  
 10  arrival_month                         36275 non-null  int64  
 11  arrival_date   

Dataset B (Hotel)

- Booking_ID = Pengenal unik dari tiap pemesanan
- no_of_adults = Jumlah orang dewasa
- no_of_children = Jumlah anak kecil
- no_of_weekend_nights = Jumlah malam akhir pekan (Sabtu atau Minggu) tamu menginap atau
memesan untuk menginap di hotel
- no_of_week_nights = Jumlah malam dalam seminggu (Senin hingga Jumat) tamu menginap atau
memesan untuk menginap di hotel
- type_of_meal_plan = Jenis paket makanan yang dipesan oleh pelanggan
- required_car_parking_space = Apakah pelanggan membutuhkan tempat parkir mobil? (0 - Tidak, 1-
Ya)
- room_type_reserved = Jenis kamar yang dipesan oleh pelanggan. Nilai-nilai tersebut dienkripsi oleh
INN Hotels
- lead_time = Jumlah hari antara tanggal pemesanan dan tanggal kedatangan
- arrival_year = Tahun tanggal kedatangan
- arrival_month = Bulan tanggal kedatangan
- arrival_date = Tanggal kedatangan
- market_segment_type = Penunjukan segmen pasar
- repeated_guest = Apakah pelanggan tersebut merupakan tamu yang pernah melakukan booking dan
juga menginap? (0 - Tidak, 1- Ya)
- no_of_previous_cancellations = Jumlah pemesanan sebelumnya yang dibatalkan oleh pelanggan
sebelum pemesanan saat ini
- no_of_previous_bookings_not_canceled = Jumlah pemesanan sebelumnya yang tidak dibatalkan
oleh pelanggan sebelum pemesanan saat ini
- avg_price_per_room = Harga rata-rata per hari pemesanan; harga kamar bersifat dinamis. (dalam
euro)
- no_of_special_requests = Jumlah total permintaan khusus yang dibuat oleh pelanggan (misalnya
lantai yang tinggi, pemandangan dari kamar, dan lain-lain.)
- booking_status (target variable) = Flag yang menunjukkan apakah pemesanan dibatalkan atau tidak.

In [7]:
df.describe()

no_of_adults  no_of_children  no_of_weekend_nights  no_of_week_nights  \
count  36275.000000    36275.000000          36275.000000       36275.000000   
mean       1.844962        0.105279              0.810724           2.204300   
std        0.518715        0.402648              0.870644           1.410905   
min        0.000000        0.000000              0.000000           0.000000   
25%        2.000000        0.000000              0.000000           1.000000   
50%        2.000000        0.000000              1.000000           2.000000   
75%        2.000000        0.000000              2.000000           3.000000   
max        4.000000       10.000000              7.000000          17.000000   

       required_car_parking_space     lead_time  arrival_year  arrival_month  \
count                35005.000000  36275.000000  36275.000000   36275.000000   
mean                     0.031281     85.232557   2017.820427       7.423653   
std                      0.174079     85.930817      0.383836       3.069894   
min                      0.000000      0.000000   2017.000000       1.000000   
25%                      0.000000     17.000000   2018.000000       5.000000   
50%                      0.000000     57.000000   2018.000000       8.000000   
75%                      0.000000    126.000000   2018.000000      10.000000   
max                      1.000000    443.000000   2018.000000      12.000000   

       arrival_date  repeated_guest  no_of_previous_cancellations  \
count  36275.000000    36275.000000                  36275.000000   
mean      15.596995        0.025637                      0.023349   
std        8.740447        0.158053                      0.368331   
min        1.000000        0.000000                      0.000000   
25%        8.000000        0.000000                      0.000000   
50%       16.000000        0.000000                      0.000000   
75%       23.000000        0.000000                      0.000000   
max       31.000000        1.000000                     13.000000   

       no_of_previous_bookings_not_canceled  avg_price_per_room  \
count                          36275.000000        34643.000000   
mean                               0.153411          103.447208   
std                                1.754171           35.158417   
min                                0.000000            0.000000   
25%                                0.000000           80.300000   
50%                                0.000000           99.450000   
75%                                0.000000          120.275000   
max                               58.000000          540.000000   

       no_of_special_requests  
count            36275.000000  
mean                 0.619655  
std                  0.786236  
min                  0.000000  
25%                  0.000000  
50%                  0.000000  
75%                  1.000000  
max                  5.000000

In [8]:
df.columns

Index(['Booking_ID', 'no_of_adults', 'no_of_children', 'no_of_weekend_nights',
       'no_of_week_nights', 'type_of_meal_plan', 'required_car_parking_space',
       'room_type_reserved', 'lead_time', 'arrival_year', 'arrival_month',
       'arrival_date', 'market_segment_type', 'repeated_guest',
       'no_of_previous_cancellations', 'no_of_previous_bookings_not_canceled',
       'avg_price_per_room', 'no_of_special_requests', 'booking_status'],
      dtype='object')

In [9]:
df.nunique()

Booking_ID                              36275
no_of_adults                                5
no_of_children                              6
no_of_weekend_nights                        8
no_of_week_nights                          18
type_of_meal_plan                           4
required_car_parking_space                  2
room_type_reserved                          7
lead_time                                 352
arrival_year                                2
arrival_month                              12
arrival_date                               31
market_segment_type                         5
repeated_guest                              2
no_of_previous_cancellations                9
no_of_previous_bookings_not_canceled       59
avg_price_per_room                       3850
no_of_special_requests                      6
booking_status                              2
dtype: int64

Kolom **Booking_ID** menjadi kolom yang paling banyak unique valuenya, karena kolom tersebut adalah sebuah pengenal/identifier, yang tentu memiliki kombinasi angka yang berbeda di setiap barisnya.

In [10]:
print(df.isnull().sum())

Booking_ID                                 0
no_of_adults                               0
no_of_children                             0
no_of_weekend_nights                       0
no_of_week_nights                          0
type_of_meal_plan                        907
required_car_parking_space              1270
room_type_reserved                         0
lead_time                                  0
arrival_year                               0
arrival_month                              0
arrival_date                               0
market_segment_type                        0
repeated_guest                             0
no_of_previous_cancellations               0
no_of_previous_bookings_not_canceled       0
avg_price_per_room                      1632
no_of_special_requests                     0
booking_status                             0
dtype: int64


Setelah ditelusuri, ternyata hanya kolom **type_of_mean_plan**, **required_car_parking_space**, dan **avg_price_per_room** mempunyai missing values. Mungkin nanti kolom kolom ini bisa didrop/diimpute menggunakan modus/median/mean.

In [11]:
df.duplicated().sum()

0

Masing-masing kolom dari dataset ini untungnya tidak memiliki nilai duplikat , sehingga tidak perlu dilakukan penghapusan nilai duplikat.

# Data Preprocessing

## Impute Missing Values

In [12]:
missingValues = df.isnull().sum()
missingPercentage = (missingValues / len(df)) * 100
dataTypes = df.dtypes
missingData = pd.DataFrame({'Amount of Missing Values': missingValues, 'Percentage of Missing Values': missingPercentage, 'Data Type': dataTypes})
missingData = missingData.sort_values(by=['Percentage of Missing Values'], ascending=False)
missingData.style

Kolom **type_of_mean_plan**, **required_car_parking_space**, dan **avg_price_per_room** secara beurutan mempunyai missing values sebesar 4,5%, 3.5%, dan 2.5% dari keseluruhan dataset. Kolom-kolom ini kita bisa handle dengan cara impute mode/median/mean dari data yang ada.

In [13]:
df['type_of_meal_plan'].value_counts()

type_of_meal_plan
Meal Plan 1     27157
Not Selected     4986
Meal Plan 2      3220
Meal Plan 3         5
Name: count, dtype: int64

Ternyata missing values pada kolom **type_of_meal_plan** dilabel dengan **Not Selected**.

In [14]:
print(df['type_of_meal_plan'].mode().sum())

Meal Plan 1


Setelah dicari tahu lebih lanjut, modus dari kolom **type_of_meal_plan** adalah **Meal Plan 1**.

In [15]:
df['type_of_meal_plan'] = df['type_of_meal_plan'].replace('Not Selected', 'Meal Plan 1')
modus = df['type_of_meal_plan'].mode()[0]
df['type_of_meal_plan'].fillna(modus, inplace=True)
df.head(10)

Booking_ID  no_of_adults  no_of_children  no_of_weekend_nights  \
0   INN00001             2               0                     1   
1   INN00002             2               0                     2   
2   INN00003             1               0                     2   
3   INN00004             2               0                     0   
4   INN00005             2               0                     1   
5   INN00006             2               0                     0   
6   INN00007             2               0                     1   
7   INN00008             2               0                     1   
8   INN00009             3               0                     0   
9   INN00010             2               0                     0   

   no_of_week_nights type_of_meal_plan  required_car_parking_space  \
0                  2       Meal Plan 1                         0.0   
1                  3       Meal Plan 1                         0.0   
2                  1       Meal Plan 1                         0.0   
3                  2       Meal Plan 1                         0.0   
4                  1       Meal Plan 1                         0.0   
5                  2       Meal Plan 2                         0.0   
6                  3       Meal Plan 1                         0.0   
7                  3       Meal Plan 1                         0.0   
8                  4       Meal Plan 1                         0.0   
9                  5       Meal Plan 1                         0.0   

  room_type_reserved  lead_time  arrival_year  arrival_month  arrival_date  \
0        Room_Type 1        224          2017             10             2   
1        Room_Type 1          5          2018             11             6   
2        Room_Type 1          1          2018              2            28   
3        Room_Type 1        211          2018              5            20   
4        Room_Type 1         48          2018              4            11   
5        Room_Type 1        346          2018              9            13   
6        Room_Type 1         34          2017             10            15   
7        Room_Type 4         83          2018             12            26   
8        Room_Type 1        121          2018              7             6   
9        Room_Type 4         44          2018             10            18   

  market_segment_type  repeated_guest  no_of_previous_cancellations  \
0             Offline               0                             0   
1              Online               0                             0   
2              Online               0                             0   
3              Online               0                             0   
4              Online               0                             0   
5              Online               0                             0   
6              Online               0                             0   
7              Online               0                             0   
8             Offline               0                             0   
9              Online               0                             0   

   no_of_previous_bookings_not_canceled  avg_price_per_room  \
0                                     0               65.00   
1                                     0              106.68   
2                                     0               60.00   
3                                     0              100.00   
4                                     0               94.50   
5                                     0              115.00   
6                                     0                 NaN   
7                                     0              105.61   
8                                     0               96.90   
9                                     0              133.44   

   no_of_special_requests booking_status  
0                       0   Not_Canceled  
1                       1   Not_Canceled  
2                 

In [16]:
df['type_of_meal_plan'].value_counts()

type_of_meal_plan
Meal Plan 1    33050
Meal Plan 2     3220
Meal Plan 3        5
Name: count, dtype: int64

Terakhir, kita impute **Meal Plan 1** yang merupakan modus dari kolom **type_of_meal_plan** ke semua nilai yang **Not Selected** dan **NaN**.

In [17]:
df['required_car_parking_space'].value_counts()

required_car_parking_space
0.0    33910
1.0     1095
Name: count, dtype: int64

In [18]:
print(df['required_car_parking_space'].mode().sum())

0.0


In [19]:
df['required_car_parking_space'].fillna(df['required_car_parking_space'].mode()[0], inplace=True)
df.head(10)

Booking_ID  no_of_adults  no_of_children  no_of_weekend_nights  \
0   INN00001             2               0                     1   
1   INN00002             2               0                     2   
2   INN00003             1               0                     2   
3   INN00004             2               0                     0   
4   INN00005             2               0                     1   
5   INN00006             2               0                     0   
6   INN00007             2               0                     1   
7   INN00008             2               0                     1   
8   INN00009             3               0                     0   
9   INN00010             2               0                     0   

   no_of_week_nights type_of_meal_plan  required_car_parking_space  \
0                  2       Meal Plan 1                         0.0   
1                  3       Meal Plan 1                         0.0   
2                  1       Meal Plan 1                         0.0   
3                  2       Meal Plan 1                         0.0   
4                  1       Meal Plan 1                         0.0   
5                  2       Meal Plan 2                         0.0   
6                  3       Meal Plan 1                         0.0   
7                  3       Meal Plan 1                         0.0   
8                  4       Meal Plan 1                         0.0   
9                  5       Meal Plan 1                         0.0   

  room_type_reserved  lead_time  arrival_year  arrival_month  arrival_date  \
0        Room_Type 1        224          2017             10             2   
1        Room_Type 1          5          2018             11             6   
2        Room_Type 1          1          2018              2            28   
3        Room_Type 1        211          2018              5            20   
4        Room_Type 1         48          2018              4            11   
5        Room_Type 1        346          2018              9            13   
6        Room_Type 1         34          2017             10            15   
7        Room_Type 4         83          2018             12            26   
8        Room_Type 1        121          2018              7             6   
9        Room_Type 4         44          2018             10            18   

  market_segment_type  repeated_guest  no_of_previous_cancellations  \
0             Offline               0                             0   
1              Online               0                             0   
2              Online               0                             0   
3              Online               0                             0   
4              Online               0                             0   
5              Online               0                             0   
6              Online               0                             0   
7              Online               0                             0   
8             Offline               0                             0   
9              Online               0                             0   

   no_of_previous_bookings_not_canceled  avg_price_per_room  \
0                                     0               65.00   
1                                     0              106.68   
2                                     0               60.00   
3                                     0              100.00   
4                                     0               94.50   
5                                     0              115.00   
6                                     0                 NaN   
7                                     0              105.61   
8                                     0               96.90   
9                                     0              133.44   

   no_of_special_requests booking_status  
0                       0   Not_Canceled  
1                       1   Not_Canceled  
2                 

Di sini, kita mengisi kolom **required_car_parking_space** dengan modusnya. Mengapa? Karena walaupun nilai-nilai dalam kolom tersebut direpresentasikan dengan angka, namun angka-angka tersebut bersifat sebagai kategori biner (angka biner). Jadi, tidak masuk akal jika kita mengisi kolom tersebut dengan median/mean, karena keseluruhan kolom tersebut hanya ingin menyatakan apakah pelanggan membutuhkan (1.0) atau tidak membutuhkan (0.0) tempat parkir.

In [20]:
df['avg_price_per_room'].value_counts

<bound method IndexOpsMixin.value_counts of 0         65.00
1        106.68
2         60.00
3        100.00
4         94.50
          ...  
36270    167.80
36271     90.95
36272     98.39
36273     94.50
36274    161.67
Name: avg_price_per_room, Length: 36275, dtype: float64>

In [21]:
df['avg_price_per_room'].fillna(df['avg_price_per_room'].median(), inplace=True)
df.head(10)

Booking_ID  no_of_adults  no_of_children  no_of_weekend_nights  \
0   INN00001             2               0                     1   
1   INN00002             2               0                     2   
2   INN00003             1               0                     2   
3   INN00004             2               0                     0   
4   INN00005             2               0                     1   
5   INN00006             2               0                     0   
6   INN00007             2               0                     1   
7   INN00008             2               0                     1   
8   INN00009             3               0                     0   
9   INN00010             2               0                     0   

   no_of_week_nights type_of_meal_plan  required_car_parking_space  \
0                  2       Meal Plan 1                         0.0   
1                  3       Meal Plan 1                         0.0   
2                  1       Meal Plan 1                         0.0   
3                  2       Meal Plan 1                         0.0   
4                  1       Meal Plan 1                         0.0   
5                  2       Meal Plan 2                         0.0   
6                  3       Meal Plan 1                         0.0   
7                  3       Meal Plan 1                         0.0   
8                  4       Meal Plan 1                         0.0   
9                  5       Meal Plan 1                         0.0   

  room_type_reserved  lead_time  arrival_year  arrival_month  arrival_date  \
0        Room_Type 1        224          2017             10             2   
1        Room_Type 1          5          2018             11             6   
2        Room_Type 1          1          2018              2            28   
3        Room_Type 1        211          2018              5            20   
4        Room_Type 1         48          2018              4            11   
5        Room_Type 1        346          2018              9            13   
6        Room_Type 1         34          2017             10            15   
7        Room_Type 4         83          2018             12            26   
8        Room_Type 1        121          2018              7             6   
9        Room_Type 4         44          2018             10            18   

  market_segment_type  repeated_guest  no_of_previous_cancellations  \
0             Offline               0                             0   
1              Online               0                             0   
2              Online               0                             0   
3              Online               0                             0   
4              Online               0                             0   
5              Online               0                             0   
6              Online               0                             0   
7              Online               0                             0   
8             Offline               0                             0   
9              Online               0                             0   

   no_of_previous_bookings_not_canceled  avg_price_per_room  \
0                                     0               65.00   
1                                     0              106.68   
2                                     0               60.00   
3                                     0              100.00   
4                                     0               94.50   
5                                     0              115.00   
6                                     0               99.45   
7                                     0              105.61   
8                                     0               96.90   
9                                     0              133.44   

   no_of_special_requests booking_status  
0                       0   Not_Canceled  
1                       1   Not_Canceled  
2                 

Untuk kolom **avg_price_per_room**, kita dapat mengisi dengan median dari keseluruhan nilai yang ada dalam kolom tersebut.

In [22]:
missingValues = df.isnull().sum()
missingPercentage = (missingValues / len(df)) * 100
dataTypes = df.dtypes
missingData = pd.DataFrame({'Amount of Missing Values': missingValues, 'Percentage of Missing Values': missingPercentage, 'Data Type': dataTypes})
missingData = missingData.sort_values(by=['Percentage of Missing Values'], ascending=False)
missingData.style

In [23]:
df['type_of_meal_plan'].value_counts()

type_of_meal_plan
Meal Plan 1    33050
Meal Plan 2     3220
Meal Plan 3        5
Name: count, dtype: int64

Dengan ini, kita berhasil meng-handle semua **missing values** yang ada pada dataset ini.

## Drop Unnecessary Column

In [24]:
df.drop(columns=['Booking_ID'], inplace=True)
df.head(10)

no_of_adults  no_of_children  no_of_weekend_nights  no_of_week_nights  \
0             2               0                     1                  2   
1             2               0                     2                  3   
2             1               0                     2                  1   
3             2               0                     0                  2   
4             2               0                     1                  1   
5             2               0                     0                  2   
6             2               0                     1                  3   
7             2               0                     1                  3   
8             3               0                     0                  4   
9             2               0                     0                  5   

  type_of_meal_plan  required_car_parking_space room_type_reserved  lead_time  \
0       Meal Plan 1                         0.0        Room_Type 1        224   
1       Meal Plan 1                         0.0        Room_Type 1          5   
2       Meal Plan 1                         0.0        Room_Type 1          1   
3       Meal Plan 1                         0.0        Room_Type 1        211   
4       Meal Plan 1                         0.0        Room_Type 1         48   
5       Meal Plan 2                         0.0        Room_Type 1        346   
6       Meal Plan 1                         0.0        Room_Type 1         34   
7       Meal Plan 1                         0.0        Room_Type 4         83   
8       Meal Plan 1                         0.0        Room_Type 1        121   
9       Meal Plan 1                         0.0        Room_Type 4         44   

   arrival_year  arrival_month  arrival_date market_segment_type  \
0          2017             10             2             Offline   
1          2018             11             6              Online   
2          2018              2            28              Online   
3          2018              5            20              Online   
4          2018              4            11              Online   
5          2018              9            13              Online   
6          2017             10            15              Online   
7          2018             12            26              Online   
8          2018              7             6             Offline   
9          2018             10            18              Online   

   repeated_guest  no_of_previous_cancellations  \
0               0                             0   
1               0                             0   
2               0                             0   
3               0                             0   
4               0                             0   
5               0                             0   
6               0                             0   
7               0                             0   
8               0                             0   
9               0                             0   

   no_of_previous_bookings_not_canceled  avg_price_per_room  \
0                                     0               65.00   
1                                     0              106.68   
2                                     0               60.00   
3                                     0              100.00   
4                                     0               94.50   
5                                     0              115.00   
6                                     0               99.45   
7                                     0              105.61   
8                                     0               96.90   
9                                     0              133.44   

   no_of_special_requests booking_status  
0                       0   Not_Canceled  
1                       1   Not_Canceled  
2                       0       Canceled  
3                       0       Canceled  
4                       0       Canceled  
5          

Drop kolom **Booking_ID** karena kolom tersebut adalah kolom pengenal yang berisi unique values, di mana kolom ini tidak diperlukan ketika ingin membuat sebuah model.

# Feature Engineering

## Encoding

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36275 entries, 0 to 36274
Data columns (total 18 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   no_of_adults                          36275 non-null  int64  
 1   no_of_children                        36275 non-null  int64  
 2   no_of_weekend_nights                  36275 non-null  int64  
 3   no_of_week_nights                     36275 non-null  int64  
 4   type_of_meal_plan                     36275 non-null  object 
 5   required_car_parking_space            36275 non-null  float64
 6   room_type_reserved                    36275 non-null  object 
 7   lead_time                             36275 non-null  int64  
 8   arrival_year                          36275 non-null  int64  
 9   arrival_month                         36275 non-null  int64  
 10  arrival_date                          36275 non-null  int64  
 11  market_segment_

In [26]:
df['market_segment_type'].value_counts()

market_segment_type
Online           23214
Offline          10528
Corporate         2017
Complementary      391
Aviation           125
Name: count, dtype: int64

In [27]:
df['type_of_meal_plan'].value_counts()

type_of_meal_plan
Meal Plan 1    33050
Meal Plan 2     3220
Meal Plan 3        5
Name: count, dtype: int64

In [28]:
df['room_type_reserved'].value_counts()

room_type_reserved
Room_Type 1    28130
Room_Type 4     6057
Room_Type 6      966
Room_Type 2      692
Room_Type 5      265
Room_Type 7      158
Room_Type 3        7
Name: count, dtype: int64

In [29]:
df['booking_status'].value_counts()

booking_status
Not_Canceled    24390
Canceled        11885
Name: count, dtype: int64

In [30]:
oneHotEnc = OneHotEncoder(sparse_output=False, drop=None)
labelEnc = LabelEncoder()

In [31]:
df['booking_status'] = labelEnc.fit_transform(df['booking_status'])
df['arrival_year'] = labelEnc.fit_transform(df['arrival_year'])

Di sini, kita menggunakan **label encoder** karena nilai dalam kolom **booking_status** dan **arrival_year** hanya ada dua nilai yang berbeda.

In [32]:
categorical_cols = ['type_of_meal_plan', 'room_type_reserved', 'market_segment_type']

encoded = oneHotEnc.fit_transform(df[categorical_cols])

encoded_df = pd.DataFrame(encoded, columns=oneHotEnc.get_feature_names_out(categorical_cols))

df = df.drop(columns=categorical_cols)
df = pd.concat([df.reset_index(drop=True), encoded_df], axis=1)

Sedangkan untuk kolom **type_of_meal_plan**, **room_type_reserved**, dan **market_segment_type**, kita dapat menggunakan karena nilai masing-masing kolom non-ordinal.

In [33]:
df.head(10)

no_of_adults  no_of_children  no_of_weekend_nights  no_of_week_nights  \
0             2               0                     1                  2   
1             2               0                     2                  3   
2             1               0                     2                  1   
3             2               0                     0                  2   
4             2               0                     1                  1   
5             2               0                     0                  2   
6             2               0                     1                  3   
7             2               0                     1                  3   
8             3               0                     0                  4   
9             2               0                     0                  5   

   required_car_parking_space  lead_time  arrival_year  arrival_month  \
0                         0.0        224             0             10   
1                         0.0          5             1             11   
2                         0.0          1             1              2   
3                         0.0        211             1              5   
4                         0.0         48             1              4   
5                         0.0        346             1              9   
6                         0.0         34             0             10   
7                         0.0         83             1             12   
8                         0.0        121             1              7   
9                         0.0         44             1             10   

   arrival_date  repeated_guest  ...  room_type_reserved_Room_Type 3  \
0             2               0  ...                             0.0   
1             6               0  ...                             0.0   
2            28               0  ...                             0.0   
3            20               0  ...                             0.0   
4            11               0  ...                             0.0   
5            13               0  ...                             0.0   
6            15               0  ...                             0.0   
7            26               0  ...                             0.0   
8             6               0  ...                             0.0   
9            18               0  ...                             0.0   

   room_type_reserved_Room_Type 4  room_type_reserved_Room_Type 5  \
0                             0.0                             0.0   
1                             0.0                             0.0   
2                             0.0                             0.0   
3                             0.0                             0.0   
4                             0.0                             0.0   
5                             0.0                             0.0   
6                             0.0                             0.0   
7                             1.0                             0.0   
8                             0.0                             0.0   
9                             1.0                             0.0   

   room_type_reserved_Room_Type 6  room_type_reserved_Room_Type 7  \
0                             0.0                             0.0   
1                             0.0                             0.0   
2                             0.0                             0.0   
3                             0.0                             0.0   
4                             0.0                             0.0   
5                             0.0                             0.0   
6                             0.0                             0.0   
7                             0.0                             0.0   
8                             0.0                             0.0   
9                             0.0                             0.0   

   market_segment_type_Aviation  market_segment_t

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36275 entries, 0 to 36274
Data columns (total 30 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   no_of_adults                          36275 non-null  int64  
 1   no_of_children                        36275 non-null  int64  
 2   no_of_weekend_nights                  36275 non-null  int64  
 3   no_of_week_nights                     36275 non-null  int64  
 4   required_car_parking_space            36275 non-null  float64
 5   lead_time                             36275 non-null  int64  
 6   arrival_year                          36275 non-null  int64  
 7   arrival_month                         36275 non-null  int64  
 8   arrival_date                          36275 non-null  int64  
 9   repeated_guest                        36275 non-null  int64  
 10  no_of_previous_cancellations          36275 non-null  int64  
 11  no_of_previous_

In [35]:
df['room_type_reserved_Room_Type 2'].head()

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: room_type_reserved_Room_Type 2, dtype: float64

# Data Splitting

In [36]:
x = df.drop(columns=['booking_status'])
y = df['booking_status']

In [37]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)
print(x_train.shape, x_test.shape)

(29020, 29) (7255, 29)


Di sini, kita melakukan splitting data **80% untuk training** dan **20% untuk testing**, di mana target variabel kita adalah kolom **booking_status**.

# Modelling

## Random Forest Model

In [38]:
RF_class = RandomForestClassifier()
RF_class.fit(x_train, y_train)

RandomForestClassifier()

In [39]:
y_pred_RF = RF_class.predict(x_test)

print("Classification Report Random Forest:\n", classification_report(y_test, y_pred_RF, target_names= ['Canceled','Not_Canceled']))

Classification Report Random Forest:
               precision    recall  f1-score   support

    Canceled       0.88      0.82      0.85      2416
Not_Canceled       0.91      0.94      0.93      4839

    accuracy                           0.90      7255
   macro avg       0.90      0.88      0.89      7255
weighted avg       0.90      0.90      0.90      7255



## XGBoost Model

In [40]:
XGB_class = XGBClassifier()
XGB_class.fit(x_train, y_train)
y_pred_XGB = XGB_class.predict(x_test)

In [41]:
print("Classification Report XGBoost:\n", classification_report(y_test, y_pred_XGB, target_names= ['Canceled','Not_Canceled']))

Classification Report XGBoost:
               precision    recall  f1-score   support

    Canceled       0.87      0.80      0.83      2416
Not_Canceled       0.90      0.94      0.92      4839

    accuracy                           0.89      7255
   macro avg       0.88      0.87      0.88      7255
weighted avg       0.89      0.89      0.89      7255



# Evaluation

**Analisis Kedua Model**

Berdasarkan hasil evaluasi classification report, model **Random Forest** memiliki performa yang sedikit lebih unggul dibanding **XGBoost**. Hal ini terlihat dari:
- Nilai accuracy yang lebih tinggi (90% vs 89%)
- Nilai macro average F1-score yang lebih baik (0.89 vs 0.88)
- F1-score untuk kedua kelas juga lebih seimbang di Random Forest

Meskipun selisihnya tipis, Random Forest menunjukkan performa yang lebih stabil dan akurat dalam menangani kedua kelas (canceled dan not_canceled). Oleh karena itu, **Random Forest** dapat dianggap sebagai model terbaik dalam kasus ini.